aws-lc-rs/src/rsa/encoding.rs

```diff
@@ -33,7 +34,126 @@ pub(in super::super) mod pkcs8 {
     }
 
     // Supports v1 and v2 encodings through a single API entry-point.
-    pub(in super::super) fn decode_der(pkcs8: &[u8]) -> Result<LcPtr<EVP_PKEY>, KeyRejected> {
-        LcPtr::try_from(pkcs8)
+    pub(in crate::rsa) fn decode_der(pkcs8: &[u8]) -> Result<LcPtr<EVP_PKEY>, KeyRejected> {
+        let mut cbs = unsafe { cbs::build_CBS(pkcs8) };
+        let key = LcPtr::new(unsafe { EVP_parse_private_key(&mut cbs) })
+            .map_err(|()| KeyRejected::invalid_encoding())?;
+        if !is_rsa_key(&key) {
+            return Err(KeyRejected::unspecified());
+        }
+        Ok(key)
+    }
+}
+
+/// [RFC 8017](https://www.rfc-editor.org/rfc/rfc8017.html)
+///
+/// PKCS #1: RSA Cryptography Specifications Version 2.2
+pub(in crate::rsa) mod rfc8017 {
+    use crate::{
+        cbs,
+        error::{KeyRejected, Unspecified},
+        ptr::{DetachableLcPtr, LcPtr},
+    };
+    use aws_lc::{
+        EVP_PKEY_assign_RSA, EVP_PKEY_new, RSA_parse_private_key, RSA_parse_public_key,
+        RSA_public_key_to_bytes, EVP_PKEY,
+    };
+    use std::ptr::null_mut;
+
+    /// DER encode a RSA public key to `RSAPublicKey` structure.
+    pub(in crate::rsa) unsafe fn encode_public_key_der(
+        pubkey: &LcPtr<EVP_PKEY>,
+    ) -> Result<Box<[u8]>, Unspecified> {
+        let mut pubkey_bytes = null_mut::<u8>();
+        let mut outlen: usize = 0;
+        if 1 != RSA_public_key_to_bytes(
+            &mut pubkey_bytes,
+            &mut outlen,
+            *pubkey.get_rsa()?.as_const(),
+        ) {
+            return Err(Unspecified);
+        }
+        let pubkey_bytes = LcPtr::new(pubkey_bytes)?;
+        let pubkey_slice = pubkey_bytes.as_slice(outlen);
+        let pubkey_vec = Vec::from(pubkey_slice);
+        Ok(pubkey_vec.into_boxed_slice())
+    }
+
+    /// Decode a DER encoded `RSAPublicKey` structure.
+    #[inline]
+    pub(in crate::rsa) fn decode_public_key_der(
+        public_key: &[u8],
+    ) -> Result<LcPtr<EVP_PKEY>, KeyRejected> {
+        let mut cbs = unsafe { cbs::build_CBS(public_key) };
+
+        let rsa = DetachableLcPtr::new(unsafe { RSA_parse_public_key(&mut cbs) })?;
```

Comment:
It returns an `Result::Err(..)` type that will be returned from this function instead. The `?` is an error propagation, so its `Result::Ok(...)` then the type inside `Ok` will be assigned to the left, otherwise this will return the error.